src/solvers/refinement/string_constraint_generator_main.cpp

```diff
@@ -319,21 +318,71 @@ void string_constraint_generatort::add_default_axioms(
   if(!created_strings.insert(s).second)
     return;
 
-  axioms.push_back(
-    s.axiom_for_length_ge(from_integer(0, s.length().type())));
+  const exprt index_zero = from_integer(0, s.length().type());
+  axioms.push_back(s.axiom_for_length_ge(index_zero));
+
   if(max_string_length!=std::numeric_limits<size_t>::max())
     axioms.push_back(s.axiom_for_length_le(max_string_length));
+}
 
-  if(force_printable_characters)
-  {
-    symbol_exprt qvar=fresh_univ_index("printable", s.length().type());
-    exprt chr=s[qvar];
-    and_exprt printable(
-      binary_relation_exprt(chr, ID_ge, from_integer(' ', chr.type())),
-      binary_relation_exprt(chr, ID_le, from_integer('~', chr.type())));
-    string_constraintt sc(qvar, s.length(), printable);
-    axioms.push_back(sc);
-  }
+/// Add constraint on characters of a string.
+///
+/// This constraint is
+/// \f$ \forall i \in [start, end), low\_char \le s[i] \le high\_char \f$
+/// \param s: a string expression
+/// \param start: index of the first character to constrain
+/// \param end: index at which we stop adding constraints
+/// \param char_set: a string of the form "<low_char>-<high_char>" where
+///        `<low_char>` and `<high_char>` are two characters, which represents
+///        the set of characters that are between `low_char` and `high_char`.
+/// \return a string expression that is linked to the argument through axioms
+///   that are added to the list
+void string_constraint_generatort::add_constraint_on_characters(
+  const array_string_exprt &s,
+  const exprt &start,
+  const exprt &end,
+  const std::string &char_set)
+{
+  // Parse char_set
+  PRECONDITION(char_set.length() == 3);
```

Comment:
Do we check this when accepting the command-line option? If not we should; it's always better to fail early on bad user input so they don't have to wait until symex completes to learn they've done it wrong

Reply:
For now, this doesn't come directly from the command-line, when `--string-printable` is set `java_object_factory` will call this primitive with the fixed input " -~" so there is no risk of bad user input.
In the future, we could allow the user to configure that and have more complex char_set (as you have in regex for instance), but this is outside the scope of this PR.  

> e.g. restricting the charset to "a-b" , insist that the string is 5 chars long and not "aaaaa", and check that you get something matching [ab]{5}

Because this cannot really be set by the user, this would require some unit test. Alternatively I could add a test with something like this:
`f(String s, char c){assert(s.find(c)=-1 || (c>= ' ' && c<= '~'));}` which ensures we cannot find a non ascii-printable character in a string.